Add aggregated topic logs

jrbourbeau opened the pull request:
This PR adds the ability to log events under a specified topic to a centralized ledger of events that lives on the scheduler. Currently the centralized ledger contains some administrative events which were previously logged using `Scheduler.log_event`. With the changes in this PR, the client and workers now also have a `log_event` method to add their own events to the ledger. Additionally, the client also has a `get_events` method which allows users to retrieve events from the ledger.

Closes https://github.com/dask/distributed/issues/4222

mrocklin commented on distributed/scheduler.py:
```diff
@@ -1298,6 +1298,8 @@ def __init__(
             maxlen=dask.config.get("distributed.scheduler.transition-log-length")
         )
+        self.events = defaultdict(lambda: deque(maxlen=100000))
+        self.event_counts = defaultdict(lambda: 0)
```
If you wanted to be ultra-concise

```suggestion
        self.event_counts = defaultdict(int)
```

jrbourbeau replied:
+1

mrocklin commented on distributed/tests/test_client.py:
```diff
@@ -6251,3 +6251,32 @@ async def test_get_task_metadata_multiple(c, s, a, b):
+    assert not await c.get_events("topic1")
+    await c.submit(foo)
+    events = await c.get_events("topic1")
+    assert len(events) == 1
+    assert events[0]["foo"] == "bar"
```
Why not something like this?

```suggestion
    assert events = ({"foo": "bar"},) 
```

Are we trying to support list/tuple ambiguity?  If so, I recommend calling `list(events)` or `tuple(events)` explicitly.

jrbourbeau replied:
There's an extra`"time"` timestamp key in each event which makes hard-coding the expected event tuple slightly more tricky. We could pop the `"time"` key off each event and then `assert` the observed tuple is as expected &mdash; which would be totally fine. I'm also okay with the current style here

mrocklin commented on distributed/tests/test_client.py:
```diff
@@ -6251,3 +6251,32 @@ async def test_get_task_metadata_multiple(c, s, a, b):
+
+    # Log an event while on the scheduler
+    def log_scheduler(dask_scheduler):
+        dask_scheduler.log_event("topic2", {"woo": "hoo"})
```
A fine contender to displace the foo-bar hegemony

mrocklin commented on distributed/scheduler.py:
```diff
@@ -3736,6 +3741,9 @@ async def feed(
                     teardown(self, state)
 
+    def log_worker_event(self, worker=None, topic=None, msg=None):
+        self.log_event(topic, msg)
```
Open question, should we include the worker address in a list of topics?

jrbourbeau replied:
I don't have strong thoughts on this. This could, for example, be useful to get all the events logged on specific workers. Though I'm inclined to hold off on that for the time being just to keep things simple initially

mrocklin commented on distributed/client.py:
```diff
@@ -3579,6 +3579,12 @@ def get_worker_logs(self, n=None, workers=None, nanny=False):
         return self.sync(self.scheduler.worker_logs, n=n, workers=workers, nanny=nanny)
 
+    def log_event(self, topic, msg):
+        return self.sync(self.scheduler.log_event, topic=topic, msg=msg)
```
Open question, what should the name of this be?  `log_event` is perfectly fine.  Do we want something that claims more space, like `log`?

jrbourbeau replied:
I have a slight preference for `log_event` as (to me) it conveys there's some additional structure beyond, for example, just logging a string in Python's normal logging system. Additionally `log` and `get_logs` are already in use elsewhere. We could certainly move things around, but I'm just not sure if it'd be worth the effort to do it right now

mrocklin reviewed (commented):
Some small comments, none of which are very important.  In general this seems good to me.  It would be useful to get some feedback from others, and maybe also include an example in a doc page somewhere.

jrbourbeau commented:
> also include an example in a doc page somewhere

I didn't see an obvious place to document this, so I added a new "Logging" doc page with information on the event logs we're adding here and a couple other logging systems already in place. This seemed reasonable to me, but I'm happy to move things around too

After thinking about it a bit more, I also updated the structure of the events we log on the scheduler to be `(timestamp, event-msg)` (this turns out to be consistent with your original proposal in https://github.com/dask/distributed/issues/4222). I like this for a couple of reasons:

1. This allows non-`dict` messages to be logged which might be preferred in some situations. For example, if logging an ML model's score I'd prefer to do:

```python
worker.log_event("my-model-score", 0.9978)
```

instead of 

```python
worker.log_event("my-model-score", {"score": 0.9978})
```

which has a bit more structure than needed.

2. This way we don't mutate the event message that is logged (previously we were adding a `"time"` key).

mrocklin commented:
This seems like it's in a good state.  Let's merge now.  We can modify in the future if necessary.
